### PR TITLE
Add String.cut

### PR DESCRIPTION
**Pull request by FkHina**
See #807

### CODE REVIEW

**Comment by thizanne** [comment]
As we used a `String.index_after_n` function, we chose to also export it since it seemed general enough to be useful by itself. We're not sure about the name, any proposal is welcome.

File: src/batString.mliv [hunk 1]
```diff
@@ -258,13 +258,18 @@ val rindex_from_opt: string -> int -> char -> int option
...
 val index_after_n :  char -> int -> string -> int
 (** [index_after_n chr n str] returns the index of the character that
-    comes immediately after the [n]-th occurrence of [chr] in [str], {b
-    starting at 1}.
+    comes immediately after the [n]-th occurrence of [chr] in [str].
+    {b n starts from 1}.
 
-    If [n] equals [0], returns [0]. If the [n]-th occurrence of [chr]
-    is the last character of [str], returns the length of [str].
+    - If [n = 0], always returns [0].
```
[the remark this answers is not in the document]
Comment (thizanne) [reply src/batString.mliv]: Tried to clarify the doc. Is it better now ?

**Comment by UnixJunkie** [comment]
A slightly more generic interface would be to allow the separator to be a string instead of a char.
Of course, a one char separator is a special case that might be optimized.

**Comment by UnixJunkie** [comment]
this is not self-documenting enough:
 +  then raise (Invalid_argument "index out of bounds") 
Cf. recent pull requests from @murmour to see good use cases of invalid_arg

**Comment by UnixJunkie** [comment]
The proposed ocamldoc is quite complex.
Saying it is inspired by unix's cut command and a few examples should be enough.

**Comment by thizanne** [comment]
- Cutting with a string as a separator is harder to define properly (what should `cut ~sep:",," ~str:"a,,,b"` do ?), to implement efficiently, and is as a matter of fact not allowed by the Unix command. We could rename `cut` to `cut_on_char` (following `split`), but I think we should keep this version anyway.
- I'm all for explicit `Invalid_argument` error messages, and we can indeed make the ones of `cut` and `index_after_n` more explicit. However I'm not fond of specifying them in the doc, since people should not rely on them in a `try` anyway.
- As for the documentation, I believe corner cases should be documented. It starts with "Similar to Unix cut" so that people may simply not read the afterpart if it's enough for them. But then they may also read them if they need the specification. I think examples would be actually longer and harder to read when they describe every corner case behaviour.

**Comment by UnixJunkie** [comment]
cut ~sep:",," ~str:"a,,,b"

since the string is processed from left to right, as the whole western world would expect, I don't see any problem.

It is OK to not specify in the doc the very precise error string that will be given in an Invalid_argument exception.

I'm fine with not changing the current ocamldoc.

I vote for cut_on_char, since we already have split_on_char and I may need cut_on_string in the near future (that will probably be named cut, because our current split is a split_on_string).

**Comment by UnixJunkie** [comment]
I think I will merge that but then rename the function to cut_on_char

**Comment by thizanne** [comment]
I'll take care of this and let you merge. Sorry for the delay.

**Comment by thizanne** [comment]
Updated to cut_on_char. I considered changing the Invalid_argument parameter to be more explicit, but since it is not actually raised by cut but by index_from and index_after_n, I gave up. This would probably need a more global work on Invalid_arg exceptions to be done properly.

**Pull request merged**